buildSrc/subprojects/parser/src/main/kotlin/org/apache/calcite/buildtools/parser/DialectGenerateTask.kt

```diff
@@ -0,0 +1,316 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.buildtools.parser
+
+import java.io.File
+import java.util.LinkedList
+import java.util.Queue
+import javax.inject.Inject
+import kotlin.collections.HashMap
+import kotlin.collections.MutableMap
+import kotlin.text.MatchResult
+import kotlin.text.Regex
+import kotlin.text.StringBuilder
+import org.gradle.api.DefaultTask
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.InputDirectory
+import org.gradle.api.tasks.TaskAction
+
+/* Gradle task that traverses the parsing directory and generates the parserImpls.ftl
+   file for the given dialect.
+ */
+open class DialectGenerateTask @Inject constructor(
+    objectFactory: ObjectFactory
+) : DefaultTask() {
+
+    private val typeAndName = "\\w+\\s+\\w+"
+    private val splitDelims = "(\\s|\n|\"|//|/\\*|\\*/|')"
+
+    private val tokenizer = Regex("((?<=$splitDelims)|(?=$splitDelims))")
+    private val declarationPattern =
+        Regex("($typeAndName\\s*\\(\\s*($typeAndName\\s*(\\,\\s*$typeAndName\\s*)*)?\\)\\s*\\:\n?)")
+    private val nameRegex = Regex("\\w+")
+
+    // Flags used when parsing.
+    private var validBrace = false
```

Comment:
Should these be declared at the class level? Couldn't this cause unwanted spillover between functions/files that we process?

Reply:
They are class level so I could create a the function determineTokenValidity() that sets them (for readability). There shouldn't be any spill over since these booleans are reset each time the curly block function is called.

Reply:
Makes sense. However, I think making them class level also harms readability because it sends a message to a reader that they have a larger scope/lifetime than they actually should.

Suggest finding another way to do this along with the potential rearchitecture of processCurlyBlock().